cleverhans/utils_eager.py

```diff
@@ -0,0 +1,280 @@
+                     dtype=X_test.dtype)
+    Y_cur = np.zeros((args.batch_size,) + Y_test.shape[1:],
+                     dtype=Y_test.dtype)
+
+    tfe = tf.contrib.eager
+    batch_x = tfe.Variable(X_test[0:args.batch_size])
+    batch_y = tfe.Variable(Y_test[0:args.batch_size])
+    for batch in range(nb_batches):
+        if batch % 100 == 0 and batch > 0:
+            _logger.debug("Batch " + str(batch))
+
+        # Must not use the `batch_indices` function here, because it
+        # repeats some examples.
+        # It's acceptable to repeat during training, but not eval.
+        start = batch * args.batch_size
+        end = min(len(X_test), start + args.batch_size)
+
+        # The last batch may be smaller than all others. This should not
+        # affect the accuarcy disproportionately.
+        cur_batch_size = end - start
+        X_cur[:cur_batch_size] = X_test[start:end]
+        Y_cur[:cur_batch_size] = Y_test[start:end]
+        tf.assign(batch_x, X_cur)
+        tf.assign(batch_y, Y_cur)
+        if attack is not None:
+            batch_adv_x = attack.generate(batch_x, **attack_args)
+            predictions = model(batch_adv_x)
+        else:
+            predictions = model(batch_x)
+        cur_corr_preds = tf.equal(tf.argmax(batch_y, axis=-1),
+                                  tf.argmax(predictions, axis=-1))
+
+        accuracy += cur_corr_preds.numpy()[:cur_batch_size].sum()
+
+    assert end >= len(X_test)
+
+    # Divide by number of examples to get final value
+    accuracy /= len(X_test)
+
+    return accuracy
+
+
+def tf_model_load(file_path=None):
+    """
+    :param file_path: path to the restored ckpt, if None is
+                      taken from FLAGS.train_dir and FLAGS.filename
+    :return:
+    """
+    saver = tf.train.Saver()
+    saver.restore(file_path)
+    return True
+
+
+def model_argmax(model, samples,):
+    """
+    Helper function that computes the current class prediction
+    :param samples: numpy array with input samples (dims must match x)
+    :return: the argmax output of predictions, i.e. the current predicted class
+    """
+    tfe = tf.contrib.eager
+    tf_samples = tfe.Variable(samples)
+    probabilities = model(tf_samples)
+
+    if samples.shape[0] == 1:
+        return tf.argmax(probabilities)
+    else:
+        return tf.argmax(probabilities, axis=1)
+
+
+def l2_batch_normalize(x, epsilon=1e-12, scope=None):
+    """
+    Helper function to normalize a batch of vectors.
+    :param x: the input
+    :param epsilon: stabilizes division
+    :return: the batch of l2 normalized vector
+    """
+    x_shape = tf.shape(x)
+    x = tf.contrib.layers.flatten(x)
+    x /= (epsilon + tf.reduce_max(tf.abs(x), 1, keep_dims=True))
+    square_sum = tf.reduce_sum(tf.square(x), 1, keep_dims=True)
+    x_inv_norm = tf.rsqrt(np.sqrt(epsilon) + square_sum)
+    x_norm = tf.multiply(x, x_inv_norm)
+    return tf.reshape(x_norm, x_shape, scope)
+
+
+def kl_with_logits(p_logits, q_logits):
```

Comment:
What is specific to eager in this function?

Reply:
SAME as previous comment.